chp07_CA/Exercise_7_01_WolframCA_randomizedrules/CA.py

```diff
@@ -0,0 +1,94 @@
+# The Nature of Code
+# Daniel Shiffman
+# http://natureofcode.com
+
+# Wolfram Cellular Automata
+
+# A class to manage the CA
+
+class CA():
+
+    w = 5
+
+    def __init__(self, r):
+        # A list to store the ruleset, for example:0,1,1,0,1,1,0,1
+        self.ruleset = r
+        # A list of 0s and 1s
+        self.cells = [0] * (width / CA.w)
+        self.restart()
+
+    # Make a random ruleset
+    def randomize(self):
+        for i in range(8):
+            self.ruleset[i] = int(random(2))
+
+    # Reset to generation 0
+    def restart(self):
+        for i in range(len(self.cells)):
+            self.cells[i] = 0
+
+        # We arbitrarily start with just the middle cell having a state of "1"
+        self.cells[len(self.cells) / 2] = 1
+        self.generation = 0
+
+    # The process of creating the new generation
+    def generate(self):
+        # First we create an empty array for the new values
+        nextgen = [0] * len(self.cells)
+        # For every spot, determine new state by examing current state, and neighbor states
+        # Ignore edges that only have one neighor
+        for i in range(1, len(self.cells) - 1):
+            left = self.cells[i - 1]   # Left neighbor state
+            me = self.cells[i]       # Current state
+            right = self.cells[i + 1]   # Right neighbor state
+            # Compute next generation state based on ruleset
+            nextgen[i] = self.rules(left, me, right)
+
+        # The current generation is the new generation
+        self.cells = nextgen
+        self.generation += 1
+
+    # This is the easy part, just draw the cells, fill 255 for '1', fill 0 for
+    # '0'
+    def display(self):
+        for i, cell in enumerate(self.cells):
+            if cell == 1:
+                fill(0)
+            else:
+                fill(255)
+            noStroke()
+            rect(i * CA.w, self.generation * CA.w, CA.w, CA.w)
+
+    # Implementing the Wolfram rules
+    # This is the concise conversion to binary way
+    # def rules(self, a, b, c):
+    #     s = str(a) + str(b) + str(c)
+    #     index = int(s, 2)
+    #     return self.ruleset[index]
+    
+    def rules(self, a, b, c):
+        ruleset = self.ruleset
+        if a == 1 and b == 1 and c == 1:
+            return ruleset[0]
```

Comment:
I don't understand. It looks like 7 returns 0 and 0 returns 7.

Anyway, if a b and c are supposed to represent bits, then

```python
return self.ruleset[a<<2 | b<<1 | c]
```

Reply:
It's quite possible I have a mistake! I tend to write things in my examples that are especially long-winded or inefficient. I'm not sure if I'm right in my thinking but the reasoning is typically to be (a) as clear as possible as to what is going on and (b) to give the reader an exercise in thinking about how to improve the code. If there's a mistake, feel free to file an issue or pull request with the original version of the code too!

Reply:
@jdf I found out the inversion by @shiffman Shiffman follows Wolfram's ruleset convention!
http://mathworld.wolfram.com/ElementaryCellularAutomaton.html
:)

Reply:
If the rules are random then just `return self.ruleset[a<<2 | b<<1 | c]`. If not, then `return self.ruleset[7 - (a<<2 | b<<1 | c)]` I guess.

Reply:
Yeah, I was testing bitwise operators around another file (so I thought), I guess I need a break, sorry.